### PR TITLE
tests: kernel: gen_isr_table: rework multi_level_bit_masks tests

### LINKED ISSUE #66365: tests: kernel: gen_isr_table: multi_level test's skip criteria
**Describe the bug**
`test_multi_level_bit_masks_sec` assumes that both `CONFIG_1ST_LEVEL_INTERRUPT_BITS` & `CONFIG_2ND_LEVEL_INTERRUPT_BITS` are LTE 10 bits.

Currently, if a DUT has `CONFIG_1ST_LEVEL_INTERRUPT_BITS` = 10 & `CONFIG_2ND_LEVEL_INTERRUPT_BITS` = 11, the test will not be skipped and would cause the CI to fail.

The above applies to `test_multi_level_bit_masks_thr` as well, which assumes that `CONFIG_1ST_LEVEL_INTERRUPT_BITS` & `CONFIG_2ND_LEVEL_INTERRUPT_BITS` are LTE 10 bits AND `CONFIG_3RD_LEVEL_INTERRUPT_BITS` is LTE 9 bits.

**To Reproduce**
1. Checkout #65500, or configure a multi-level IRQ board to have bits more than those specified in `test_multi_level_bit_masks_sec` or `test_multi_level_bit_masks_thr`
2. `west build -b <board> -p auto -t run -T zephyr/tests/kernel/gen_isr_table/arch.interrupt.gen_isr_table.riscv_no_direct`
3. See error

**Expected behavior**
Test should pass

**Impact**
Blocking #65500, which has 11 bits for `CONFIG_2ND_LEVEL_INTERRUPT_BITS`

**Environment (please complete the following information):**
 - OS: Linux
 - Toolchain: Zephyr SDK
 - Commit SHA or Version used: 0bf4efc78080ceefaad7becf0d01c8d7fb2a8119

### PR DESCRIPTION
**Pull request by ycsin**
Reworked and combined the multi-level interrupt bit masks tests
to work on any configuration of bits in every level.

Updated the bits configuration in the testcase to use
non-symetric numbers so that shifting a level with the wrong
number of bits will certainly cause the test to fail.

Fixes #66365

### CODE REVIEW

**Comment by ycsin** [comment]
Added `Hotfix` tag as it is failing in #65500

**Review by peter-mitsis — DISMISSED** [review]

**Review by cfriedt — CHANGES_REQUESTED** [review]
Personally, I think this test would be significantly better off without the magic numbers.

Why even bother skipping tests if it's just a matter of using the correct bitmask?

File: tests/kernel/gen_isr_table/src/main.c [hunk 1]
```diff
@@ -385,7 +385,7 @@ static void *gen_isr_table_setup(void)
 #ifdef CONFIG_MULTI_LEVEL_INTERRUPTS
 ZTEST(gen_isr_table, test_multi_level_bit_masks_sec)
 {
-#if CONFIG_1ST_LEVEL_INTERRUPT_BITS < 10 && CONFIG_2ND_LEVEL_INTERRUPT_BITS < 10
+#if CONFIG_1ST_LEVEL_INTERRUPT_BITS < 10 || CONFIG_2ND_LEVEL_INTERRUPT_BITS < 10
```
Comment (cfriedt) [inline tests/kernel/gen_isr_table/src/main.c:388]: On this line itself  is the magic number 10. I can't comment on unchanged lines, but directly below this, there is the magic number 0x400. Further down, there are 0x40 and 0x800, 0x1, ..

Is there a way to make these tests pass for any valid selection of `CONFIG_1ST_LEVEL_INTERRUPT_BITS`, `CONFIG_2ND_LEVEL_INTERRUPT_BITS`, and / or `CONFIG_3RD_LEVEL_INTERRUPT_BITS`?

File: tests/kernel/gen_isr_table/src/main.c [hunk 2]
```diff
@@ -422,7 +422,8 @@ ZTEST(gen_isr_table, test_multi_level_bit_masks_sec)
 #ifdef CONFIG_3RD_LEVEL_INTERRUPTS
 ZTEST(gen_isr_table, test_multi_level_bit_masks_thr)
 {
-#if CONFIG_2ND_LEVEL_INTERRUPT_BITS < 10 && CONFIG_3RD_LEVEL_INTERRUPT_BITS < 9
+#if CONFIG_1ST_LEVEL_INTERRUPT_BITS < 10 || CONFIG_2ND_LEVEL_INTERRUPT_BITS < 10 ||                \
+	CONFIG_3RD_LEVEL_INTERRUPT_BITS < 9
```
[the remark this answers is not in the document]
Comment (cfriedt) [reply tests/kernel/gen_isr_table/src/main.c:426]: In terms of boundary checks, I would suggest checking for 0, 1, N-1, and then maybe throw in a "happy path" value as well (some number M such that 1 < M < N - 1)

File: tests/kernel/gen_isr_table/src/main.c (hunk 1 above)
Comment (ycsin) [reply tests/kernel/gen_isr_table/src/main.c]: I've updated the test to work regardless of the number of bits in each level

**Review by cfriedt — DISMISSED** [review]

File: tests/kernel/gen_isr_table/src/main.c [hunk 3]
```diff
@@ -383,96 +383,103 @@ static void *gen_isr_table_setup(void)
...
-	zassert_equal(1, ret_irq);
+	const uint32_t l2_shift = CONFIG_1ST_LEVEL_INTERRUPT_BITS;
+	const uint32_t l3_shift = CONFIG_1ST_LEVEL_INTERRUPT_BITS + CONFIG_2ND_LEVEL_INTERRUPT_BITS;
+	const uint32_t hwirq1 = irq1;
+	const uint32_t hwirq2 = irq2 - 1;
+	const uint32_t hwirq3 = irq3 - 1;
+	const bool has_l3 = irq3 > 0;
+	const bool has_l2 = irq2 > 0;
+	const uint32_t level = has_l3 ? 3 : has_l2 ? 2 : 1;
+	const uint32_t irqn = (irq3 << l3_shift) | (irq2 << l2_shift) | irq1;
```
Comment (ycsin) [inline tests/kernel/gen_isr_table/src/main.c:396]: Force pushed to update this line from
```c
	const uint32_t irqn =
		(has_l3 ? irq3 << l3_shift : 0) | (has_l2 ? irq2 << l2_shift : 0) | irq1;
```

**Review by cfriedt — APPROVED** [review]

**Review by jgl-meta — APPROVED** [review]

**Review by peter-mitsis — APPROVED** [review]

**Pull request merged**